### DISCUSSION TITLE
Ignore `null` when used as primary key

### DISCUSSION THREAD

**Discussion by MoltenCoffee** [directus/directus#6448] [Feature Requests]
When using the Directus SDK, if you make a delete request, like this (notice one of the `id`s being null):

```js
await directus.items["something"].deleteMany([ null, 2, 4, 6, 38 ]);
```

A `TransportError` will be thrown, specifically because of a `403 Forbidden` HTTP response. I assume this will also be the case when using the API, but haven't been able to confirm this.

It seems to me this is undesirable. I'd personally prefer the SDK or API (not entirely sure where this error originates) to ignore nulls altogether, at least in arrays for the `xxxMany()` actions.
If it's preferred to error on nulls, I think a http `400 Bad Request` is more fitting, or better even something more verbose.

Environment:

- Node 16.0.0
- SDK 9.0.0-rc.76
- Directus 9.0.0-rc.76

Complete error:

```json
{
  "parent": {
    "message": "Request failed with status code 403",
    "name": "Error",
    "stack": "Error: Request failed with status code 403\n    at createError (J:\\XXXXX\\yyyyyy\\node_modules\\axios\\lib\\core\\createError.js:16:15)\n    at settle (J:\\XXXXX\\yyyyyy\\node_modules\\axios\\lib\\core\\settle.js:17:12)\n    at IncomingMessage.handleStreamEnd (J:\\XXXXX\\yyyyyy\\node_modules\\axios\\lib\\adapters\\http.js:260:11)\n    at IncomingMessage.emit (node:events:377:35)\n    at endReadableNT (node:internal/streams/readable:1312:12)\n    at processTicksAndRejections (node:internal/process/task_queues:83:21)",
    "config": {
      "url": "/items/newsletter_subs",
      "method": "delete",
      "data": "[null,7,8]",
      "headers": {
        "Accept": "application/json, text/plain, */*",
        "Authorization": "Bearer ZZZZZZZZZZZ",
        "Content-Type": "application/json;charset=utf-8",
        "User-Agent": "axios/0.21.1",
        "Content-Length": 10
      },
      "baseURL": "http://localhost:2204",
      "transformRequest": [null],
      "transformResponse": [null],
      "timeout": 0,
      "withCredentials": true,
      "xsrfCookieName": "XSRF-TOKEN",
      "xsrfHeaderName": "X-XSRF-TOKEN",
      "maxContentLength": -1,
      "maxBodyLength": -1
    }
  },
  "response": {
    "raw": {
      "errors": [
        {
          "message": "You don't have permission to access this.",
          "extensions": {
            "code": "FORBIDDEN",
            "stack": "Error: You don't have permission to access this.\n    at AuthorizationService.checkAccess (J:\\XXXXX\\xxxxx\\node_modules\\directus\\dist\\services\\authorization.js:231:23)\n    at runMicrotasks (<anonymous>)\n    at processTicksAndRejections (node:internal/process/task_queues:96:5)\n    at async ItemsService.deleteMany (J:\\XXXXX\\xxxxx\\node_modules\\directus\\dist\\services\\items.js:501:13)\n    at async J:\\XXXXX\\xxxxx\\node_modules\\directus\\dist\\controllers\\items.js:160:9"
          }
        }
      ]
    },
    "status": 403,
    "statusText": "Forbidden",
    "headers": {
      "x-powered-by": "Directus",
      "access-control-allow-origin": "*",
      "access-control-allow-credentials": "true",
      "access-control-expose-headers": "Content-Range",
      "content-type": "application/json; charset=utf-8",
      "content-length": "661",
      "etag": "W/\"295-VsHiqAVimROoMBGsUHBV4/Ctrqg\"",
      "date": "Sat, 19 Jun 2021 10:34:08 GMT",
      "connection": "close"
    },
    "errors": [
      {
        "message": "You don't have permission to access this.",
        "extensions": {
          "code": "FORBIDDEN",
          "stack": "Error: You don't have permission to access this.\n    at AuthorizationService.checkAccess (J:\\XXXXX\\xxxxx\\node_modules\\directus\\dist\\services\\authorization.js:231:23)\n    at runMicrotasks (<anonymous>)\n    at processTicksAndRejections (node:internal/process/task_queues:96:5)\n    at async ItemsService.deleteMany (J:\\XXXXX\\xxxxx\\node_modules\\directus\\dist\\services\\items.js:501:13)\n    at async J:\\XXXXX\\xxxxx\\node_modules\\directus\\dist\\controllers\\items.js:160:9"
        }
      }
    ]
  },
  "errors": [
    {
      "message": "You don't have permission to access this.",
      "extensions": {
        "code": "FORBIDDEN",
        "stack": "Error: You don't have permission to access this.\n    at AuthorizationService.checkAccess (J:\\XXXXX\\xxxxx\\node_modules\\directus\\dist\\services\\authorization.js:231:23)\n    at runMicrotasks (<anonymous>)\n    at processTicksAndRejections (node:internal/process/task_queues:96:5)\n    at async ItemsService.deleteMany (J:\\XXXXX\\xxxxx\\node_modules\\directus\\dist\\services\\items.js:501:13)\n    at async J:\\XXXXX\\xxxxx\\node_modules\\directus\\dist\\controllers\\items.js:160:9"
      }
    }
  ]
}
```

**Comment by rijkvanzanten** [comment]
> If it's preferred to error on nulls, I think a http 400 Bad Request is more fitting, or better even something more verbose.

This has come up a couple times before. Right now, everything (404s, invalid IDs, etc) are returned as the same 403 forbidden, so there's no way somebody outside the project can scrape the existing record IDs. I'd like to add a "error level" environment variable that basically lets you choose how "friendly" you want the API to respond when something is happening 🤔

**Reply by MoltenCoffee** [reply]
Ah! That explains a lot, and actually sounds like a sensible decision. If scraping IDs is a problem, I'd say -most of the time- that this is a problem in database design and/or role permissions, but it seems like a sensible default. Thanks for the explanation!

I'd still argue that in `deleteMany` and `updateMany` requests, `null` values should be ignored.

**Reply by rijkvanzanten** [reply]
> I'd still argue that in deleteMany and updateMany requests, null values should be ignored.

I agree from a SDK usage perspective, but from the API input itself, I think it should throw errors on illegal input, instead of trying to silently "fix" it 🤔 Issues that come from that sort of API magic are super hard to debug

**Comment by jaads** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
